File: modules/generate/test/test_simulated_annealing_generator.cpp
/* ********************************************************************
 * Copyright (C) 2020 Pablo Hernandez-Cerdan.
 *
 * This file is part of SGEXT: http://github.com/phcerdan/sgext.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 * *******************************************************************/

#include "rng.hpp"
#include "gmock/gmock.h"

#include "simulated_annealing_generator.hpp"
#include "spatial_graph.hpp"
#include <boost/graph/graphviz.hpp>
#include <iostream>

struct SimulatedAnnealingGeneratorFixture : public ::testing::Test {

    void SetUp() override {
        RNG::engine().seed(10);
        graph_ = SG::GraphType(4);
        graph_[0].pos = {{0.25, 0.25, 0.25}};
        graph_[1].pos = {{0.5, 0.25, 0.25}};
        graph_[2].pos = {{0.25, 0.5, 0.25}};
        graph_[3].pos = {{0.25, 0.9, 0.25}};
        boost::add_edge(0, 1, SG::SpatialEdge(), graph_);
        boost::add_edge(0, 2, SG::SpatialEdge(), graph_);
        boost::add_edge(2, 3, SG::SpatialEdge(), graph_);
    }
    SG::GraphType graph_;
};

TEST_F(SimulatedAnnealingGeneratorFixture, default_constructor) {
    std::cout << "simulated_annealing works" << std::endl;
    auto gen = SG::simulated_annealing_generator();
}
TEST_F(SimulatedAnnealingGeneratorFixture, cosines_works) {
    auto gen = SG::simulated_annealing_generator(graph_);
    const size_t num_bins_ete_distances = 100;
    const size_t num_bins_cosines = 100;
    gen.init_histograms(num_bins_ete_distances, num_bins_cosines);
    gen.histo_cosines_.PrintBreaksAndCounts(std::cout);
    EXPECT_EQ(gen.histo_cosines_
                      .counts[gen.histo_cosines_.IndexFromValue(0.00000001)],
              1);
    EXPECT_EQ(gen.histo_cosines_
                      .counts[gen.histo_cosines_.IndexFromValue(-0.999)],
              1);
}

TEST_F(SimulatedAnnealingGeneratorFixture, update_step_move_node) {
    auto gen = SG::simulated_annealing_generator(graph_);
    const size_t num_bins_ete_distances = 100;
    const size_t num_bins_cosines = 100;
    gen.init_histograms(num_bins_ete_distances, num_bins_cosines);
    gen.histo_cosines_.PrintBreaksAndCounts(std::cout);
    gen.print(std::cout);
    gen.transition_params.UPDATE_STEP_MOVE_NODE_PROBABILITY = 1.0;
    gen.transition_params.MAX_ENGINE_ITERATIONS = 2;
    gen.engine();
    gen.print(std::cout);
}

// TEST_F(SimulatedAnnealingGeneratorFixture, update_step_swap_edges) {
//     auto gen = SG::simulated_annealing_generator(10);
//     gen.histo_cosines_.PrintBreaksAndCounts(std::cout);
//     gen.print(std::cout);
//     gen.transition_params.UPDATE_STEP_MOVE_NODE_PROBABILITY = 0.0;
//     gen.transition_params.MAX_ENGINE_ITERATIONS = 100;
//     gen.engine();
//     gen.print(std::cout);
// }

TEST_F(SimulatedAnnealingGeneratorFixture, num_vertices_constructor_works) {
    std::cout << "simulated_annealing works" << std::endl;
    auto gen = SG::simulated_annealing_generator(100);
    const size_t num_bins_ete_distances = 100;
    const size_t num_bins_cosines = 100;
    gen.init_histograms(num_bins_ete_distances, num_bins_cosines);
    std::cout << "Engine starting in test" << std::endl;
    gen.transition_params.UPDATE_STEP_MOVE_NODE_PROBABILITY = 0.5;
    gen.transition_params.MAX_ENGINE_ITERATIONS = 1000;
    gen.print(std::cout);
    gen.engine();
    gen.print_histo_and_target_distribution_ete_distances(std::cout);
    gen.print_histo_and_target_distribution_cosines(std::cout);
    gen.print(std::cout);
}

TEST_F(SimulatedAnnealingGeneratorFixture, config_tree_constructor_works) {
    std::cout << "simulated_annealing config tree works" << std::endl;
    auto tree = SG::simulated_annealing_generator_config_tree();
    tree.transition_params.MAX_ENGINE_ITERATIONS = 1000;
    tree.transition_params.ENERGY_CONVERGENCE = 100;
    tree.transition_params.UPDATE_STEP_MOVE_NODE_PROBABILITY = 0.5;
    tree.ete_distance_params.num_bins = 100;
    tree.cosine_params.num_bins = 100;
    auto gen = SG::simulated_annealing_generator(tree);
    gen.engine();
    gen.print(std::cout);
}

File: wrap/test/generate/test_simulated_annealing_generator.py
"""
Copyright (C) 2020 Pablo Hernande-Cerdan.

This file is part of SGEXT: http://github.com/phcerdan/sgext.

This file may be used under the terms of the GNU General Public License
version 3 as published by the Free Software Foundation and appearing in
the file LICENSE.GPL included in the packaging of this file.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""

import _sgext.generate as generate
import unittest

# The order of the test is important.
# They are run alphabetically, ie
# test_a_x will run before test_b_y
# we write the files before reading them

class TestSimulatedAnnealingAConfigTree(unittest.TestCase):
    def setUp(self):
        self.parameters_file = "fixture_sa_parameters.txt"

    def test_a_config_tree_save(self):
        config_tree = generate.simulated_annealing_generator_config_tree()
        config_tree.degree_params.max_degree = 20
        config_tree.physical_scaling_params.num_vertices = 100
        config_tree.cosine_params.num_bins = 100
        config_tree.ete_distance_params.num_bins = 100
        config_tree.transition_params.MAX_ENGINE_ITERATIONS = 1000
        config_tree.save(self.parameters_file)

    def test_b_config_tree_load(self):
        config_tree = generate.simulated_annealing_generator_config_tree()
        config_tree.load(self.parameters_file)
        self.assertEqual(config_tree.degree_params.max_degree, 20)

class TestSimulatedAnnealingBParameters(unittest.TestCase):

    def test_a_domain_parameters(self):
        par = generate.domain_parameters()
        print(par)

    def test_b_physical_scaling_parameters(self):
        par = generate.physical_scaling_parameters()
        print(par)

    def test_c_transition_parameters(self):
        par = generate.transition_parameters()
        print(par)

    def test_d_degree_distribution_parameters(self):
        par = generate.degree_distribution_parameters()
        print(par)

    def test_e_end_to_end_distances_distribution_parameters(self):
        par = generate.end_to_end_distances_distribution_parameters()
        print(par)

    def test_f_cosine_directors_distribution_parameters(self):
        par = generate.cosine_directors_distribution_parameters()
        print(par)


class TestSimulatedAnnealingC(unittest.TestCase):
    def setUp(self):
        self.parameters_file = "fixture_sa_parameters.txt"
        self.output_file = "fixture_sa_parameters_output.txt"

    def test_simulated_annealing(self):
        gen = generate.simulated_annealing_generator(self.parameters_file)
        gen.engine()
        gen.save_parameters_to_file(self.output_file)

if __name__ == '__main__':
    unittest.main()
